tik/tikSwap/tikSwap.cpp

```diff
@@ -88,196 +77,317 @@ int main(int argc, char *argv[])
     {
         tikBitcode = parseIRFile(InputFile, tikSmerror, tikContext);
     }
+    // if the file was not found
     catch (exception &e)
     {
-        std::cerr << "Couldn't open input bitcode file: " << InputFile << "\n";
-        std::cerr << e.what() << '\n';
-        spdlog::critical("Failed to open source bitcode: " + InputFile);
+        spdlog::critical("Failed to open tik bitcode file: " + InputFile);
         return EXIT_FAILURE;
     }
+    // if the parsing failed
     if (tikBitcode == nullptr)
     {
-        std::cerr << "Couldn't open input bitcode file: " << InputFile << "\n";
-        spdlog::critical("Failed to open source bitcode: " + InputFile);
+        string err;
+        raw_string_ostream rso(err);
+        tikSmerror.print(InputFile.data(), rso);
+        spdlog::critical("Couldn't parse tik bitcode:\n" + rso.str());
         return EXIT_FAILURE;
     }
-    Module *tikModule = tikBitcode.get();
 
     // grab all kernel functions in the tik bitcode and construct objects from them
-    vector<Function *> kernelFuncs;
-    for (auto &func : *tikModule)
-    {
-        string funcName = func.getName();
-        if (funcName == "K" + to_string(kernelFuncs.size()))
-        {
-            kernelFuncs.push_back(tikModule->getFunction(funcName));
-        }
-    }
     vector<TikKernel *> kernels;
-    for (auto func : kernelFuncs)
+    for (auto &func : *tikBitcode)
     {
-        auto kern = new TikKernel(func);
-        if (kern->Valid)
-        {
-            kernels.push_back(kern);
-        }
-        else
+        if (func.hasMetadata("Boundaries"))
         {
-            delete kern;
+            auto kern = new TikKernel(tikBitcode->getFunction(func.getName()));
```

Comment:
Please avoid new and delete like the plague. Use shared_ptr or unique_ptr depending on your context.

shared_ptr is probably the right answer here.

Reply:
gave that a try initially. Can't remember the problem I had with it but I think it had to do with segfaulting when trying to access the shared pointer in the kernels container later. Didn't know how to fix it so reverted to this

Reply:
The garbage collector disposes of the pointer as soon as it passes out of scope. If you want to maintain it put it in a vector in the lifetime of the object you are trying to store.